### DISCUSSION TITLE
[QST] deserializeStream() will acquire gpu semaphore，but why serializeStream() do not release the gpu semaphore explicitly ?

### DISCUSSION THREAD

**Discussion by JustPlay** [NVIDIA/spark-rapids#5386] [General] [answered]
https://github.com/NVIDIA/spark-rapids/blob/1a2b17e02f8b39cc34aadbcc43cbfff6a7724e87/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuColumnarBatchSerializer.scala#L49

https://github.com/NVIDIA/spark-rapids/blob/1a2b17e02f8b39cc34aadbcc43cbfff6a7724e87/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuColumnarBatchSerializer.scala#L132

deserializeStream() will acquire gpu semaphore，but why serializeStream() do not release the gpu semaphore explicitly ?

I think release the gpu semaphore explicitly will help increase cpu concurrenty for loading the next batches.

@jlowe @revans2

**Comment by revans2** [comment, marked as answer]
That is because it is already released.

https://github.com/NVIDIA/spark-rapids/blob/1a2b17e02f8b39cc34aadbcc43cbfff6a7724e87/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala#L85

We should document it better and if you want to turn this into a documentation bug I would be happy to fix it. The shuffle code went through a number of iterations to get it to be performant.  Sadly there is a lot of coupling between the serializer and the shuffle code and we need to document it better.

https://github.com/NVIDIA/spark-rapids/blob/1a2b17e02f8b39cc34aadbcc43cbfff6a7724e87/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuColumnarBatchSerializer.scala#L73-L76

is the only code in the serializer that works with data still on the GPU.  That code is likely dead code.  I believe I left it in there just as a precaution, but it should be cleaned up a lot.

I hope that this helps.
